src/cp/cp_control_point.c

```diff
@@ -0,0 +1,873 @@
+ *     disabled or not triggered, return NULL. If triggered, return the control point data. The data
+ *     does not need to be released.
+ *
+ * @param session The session. @param id The per connection control point's ID.
+ */
+WT_CONTROL_POINT *
+__wt_session_control_point_test_and_trigger(WT_SESSION_IMPL *session, wt_control_point_id_t id)
+{
+    WT_CONTROL_POINT *data;
+    WT_CONTROL_POINT_REGISTRY *cp_registry;
+    size_t new_crossing_count;
+    size_t new_trigger_count;
+    bool triggered;
+
+    if (WT_UNLIKELY(id >= SESSION_CONTROL_POINTS_SIZE)) {
+        __wt_verbose_error(session, WT_VERB_CONTROL_POINT,
+          "ERROR: id(%" PRId32 ") >= SESSION_CONTROL_POINTS_SIZE(%" PRId32 ")", id,
+          SESSION_CONTROL_POINTS_SIZE);
+        return (NULL);
+    }
+    if (WT_UNLIKELY(session->control_points == NULL)) {
+        __wt_verbose_warning(
+          session, WT_VERB_CONTROL_POINT, "control_points is NULL: id=%" PRId32, id);
+        return (NULL);
+    }
+    cp_registry = &(session->control_points[id]);
+
+    data = cp_registry->data;
+    if (data == NULL) {
+        /* Disabled. */
+        __wt_verbose_debug5(session, WT_VERB_CONTROL_POINT, "Is disabled: id=%" PRId32, id);
+        return (NULL);
+    }
+
+    triggered = cp_registry->pred ? cp_registry->pred(session, cp_registry, data) : true;
+    new_crossing_count = ++(cp_registry->crossing_count);
+    if (triggered) {
+        new_trigger_count = ++(cp_registry->trigger_count);
+        __wt_verbose_debug1(session, WT_VERB_CONTROL_POINT,
+          "Triggered: id=%" PRId32 ", crossing_count=%" PRIu64 ", trigger_count=%" PRIu64, id,
+          (uint64_t)new_crossing_count, (uint64_t)new_trigger_count);
+    } else {
+        __wt_verbose_debug3(session, WT_VERB_CONTROL_POINT,
+          "Not Triggered: id=%" PRId32 ", crossing_count=%" PRIu64 ", trigger_count=%" PRIu64, id,
+          (uint64_t)new_crossing_count, (uint64_t)cp_registry->trigger_count);
+        /* Not triggered. */
+        data = NULL;
+    }
+    return (data);
+}
+
+/*
+ * Per connection control point initialization.
+ */
+/* From examples/ex_control_points.c */
+/*
+ * Per connection control point "Main Start Printing".
+ */
+/* Per connection control point data type. */
+struct __wt_conn_control_point_data_main_start_printing {
+    WT_CONTROL_POINT iface;
+    WT_CONTROL_POINT_ACTION_WAIT_FOR_TRIGGER action_data;
+};
+
+/* Per connection control point init function. */
+/*
+ * __wt_conn_control_point_init_main_start_printing --
+ *     The per connection control point initialization function for control point "Main Start
+ *     Printing".
+ *
+ * @param session The session. @param cp_registry The per connection control point's control point
+ *     registry. @param cfg Configuration strings.
+ */
+WT_CONTROL_POINT *
+__wt_conn_control_point_init_main_start_printing(WT_SESSION_IMPL *session, const char **cfg)
```

Comment:
If I've understood correctly, these functions are for the control point example. If so, would they be part of the example code, rather than here?

Reply:
- Most control points will be in WiredTiger, not test programs.
- It comes down to how complicated do we want it to be to write tests using control points? Initialization and shutdown of control points is handled inside WiredTiger. To do that all control points need to be implemented inside WiredTiger even if they are used only outside WiredTiger. If we want to move code for control points not used inside WiredTiger into a test program we need to duplicate control point initialization and shutdown in the test program, have four control point ID types instead of two (inside WT per connection, inside WT per session, inside test program per connection, and inside test program per session), have four define control point macros instead of two (the same four as before), and have four call control point macros instead of four (the same four as before).